File: src/lib/watermark/darkroom-proof.ts
import { normalizeModel } from "@/lib/exif/brand";
import type { FrameParams, TemplateConfig } from "@/stores/types";
import {
  cleanDisplayText,
  formatTakenAt,
  getPreviewFontFamily,
  sanitizeCustomLines,
} from "./classic-bottom";
import {
  dprFor,
  fitPhoto,
  getCanvasRatio,
  paramsLine,
  type ShowcasePhoto,
} from "./renderer-utils";

const PAPER_BLACK = "#0d0d0d";
const PAPER_WHITE = "#f5efe1";
const PAPER_INK = "#1a1a1a";
const PAPER_MUTED = "#666058";
const PROOF_RED = "#b22222";

/**
 * Darkroom proof print. Black canvas with a creamy white photo paper that
 * has the photo printed inside; below the photo, a "PROOF" stamp in red
 * and the EXIF in old-typewriter style metadata.
 */
export function drawDarkroomProofPreview(
  canvas: HTMLCanvasElement,
  image: HTMLImageElement,
  _logo: HTMLImageElement | null,
  photo: ShowcasePhoto,
  frameParams: FrameParams,
  config: TemplateConfig,
) {
  const ctx = canvas.getContext("2d");
  if (!ctx) return;

  const dpr = dprFor(canvas);
  const baseWidth = 900;
  const ratio = getCanvasRatio(frameParams.canvasRatio, frameParams.canvasOrientation);
  const canvasW = baseWidth;
  const canvasH = canvasW / ratio;
  const watermarkActive = config.showWatermark ?? true;

  // Photo paper layout — wide white border around the photo, like darkroom prints.
  const paperPadX = canvasW * 0.06;
  const paperPadTop = canvasH * 0.06;
  const captionH = watermarkActive ? Math.max(76, canvasH * 0.14) : canvasH * 0.06;
  const paperX = canvasW * 0.04;
  const paperY = canvasH * 0.04;
  const paperW = canvasW - paperX * 2;
  const paperH = canvasH - paperY * 2;
  const photoArea = {
    x: paperX + paperPadX,
    y: paperY + paperPadTop,
    w: paperW - paperPadX * 2,
    h: paperH - paperPadTop - captionH,
  };
  const placed = fitPhoto(photoArea, photo.width, photo.height);

  canvas.width = Math.round(canvasW * dpr);
  canvas.height = Math.round(canvasH * dpr);
  canvas.style.aspectRatio = `${canvasW} / ${canvasH}`;
  ctx.setTransform(dpr, 0, 0, dpr, 0, 0);
  ctx.clearRect(0, 0, canvasW, canvasH);

  ctx.fillStyle = PAPER_BLACK;
  ctx.fillRect(0, 0, canvasW, canvasH);

  // White photo paper.
  ctx.fillStyle = PAPER_WHITE;
  ctx.fillRect(paperX, paperY, paperW, paperH);

  ctx.drawImage(image, placed.x, placed.y, placed.w, placed.h);

  if (!watermarkActive) return;

  const fontFamily = getPreviewFontFamily(frameParams.fontFamily);
  const captionTop = placed.y + placed.h + canvasH * 0.025;

  // PROOF stamp (rotated -8°), positioned over the bottom-left.
  ctx.save();
  const stampX = placed.x;
  const stampY = captionTop + Math.max(28, canvasH * 0.045);
  ctx.translate(stampX, stampY);
  ctx.rotate(-0.14);
  const stampSize = Math.max(22, canvasW * 0.038);
  ctx.font = `900 ${stampSize}px ${fontFamily}`;
  ctx.fillStyle = PROOF_RED;
  ctx.strokeStyle = PROOF_RED;
  ctx.lineWidth = 2;
  ctx.textBaseline = "alphabetic";
  ctx.textAlign = "left";
  const stampText = "PROOF";
  const stampW = ctx.measureText(stampText).width;
  ctx.strokeRect(-6, -stampSize * 0.95, stampW + 12, stampSize * 1.25);
  ctx.fillText(stampText, 0, 0);
  ctx.restore();

  const camera = config.showCamera
    ? normalizeModel(photo.exif.camera.make, photo.exif.camera.model)
    : "";
  const lens = config.showLens ? cleanDisplayText(photo.exif.lens) : "";
  const customLines = sanitizeCustomLines(config.customLines);

  // Right column — typewriter feel using mono-leaning font + monospaced fallback.
  const rightX = placed.x + placed.w;
  const meta = [
    config.showParams ? paramsLine(photo.exif, "  ") : "",
    [camera, lens].filter(Boolean).join("  ·  "),
    config.showDate ? formatTakenAt(photo.exif.takenAt, config.dateFormat) : "",
    customLines[0] || "",
  ].filter(Boolean);

  ctx.textBaseline = "alphabetic";
  ctx.textAlign = "right";
  const lineSize = Math.max(11, frameParams.fontSize);
  meta.forEach((line, i) => {
    ctx.font = `${i === 0 ? "700" : "500"} ${lineSize}px "Courier New", "Menlo", ${fontFamily}`;
    ctx.fillStyle = i === 0 ? PAPER_INK : PAPER_MUTED;
    ctx.fillText(line, rightX, captionTop + i * lineSize * 1.55 + lineSize);
  });
}

File: src-tauri/src/render/templates/darkroom_proof.rs
use image::{imageops::overlay, DynamicImage, Rgba, RgbaImage};

use crate::render::classic_bottom::{
    build_params_line, clean_display_text, compute_canvas_size, fill_rect, fit_photo_in_area,
    format_camera, format_date, resize_photo, ExportExif, ExportFrameParams, ExportTemplateConfig,
};
use crate::render::layout_spec::watermark_layout_spec;
use crate::render::text::TextRenderer;

const PAPER_BLACK: Rgba<u8> = Rgba([13, 13, 13, 255]);
const PAPER_WHITE: Rgba<u8> = Rgba([245, 239, 225, 255]);
const PAPER_INK: Rgba<u8> = Rgba([26, 26, 26, 255]);
const PAPER_MUTED: Rgba<u8> = Rgba([102, 96, 88, 255]);
const PROOF_RED: Rgba<u8> = Rgba([178, 34, 34, 255]);

pub(crate) fn compose_darkroom_proof(
    source: DynamicImage,
    frame: &ExportFrameParams,
    exif: &ExportExif,
    config: &ExportTemplateConfig,
    text: Option<&TextRenderer>,
) -> RgbaImage {
    let src_w = source.width();
    let src_h = source.height();
    let (canvas_w, canvas_h) = compute_canvas_size(src_w, src_h, frame);
    let rs = src_w as f32 / 900.0;

    let paper_x = (canvas_w as f32 * 0.04).round() as u32;
    let paper_y = (canvas_h as f32 * 0.04).round() as u32;
    let paper_w = canvas_w.saturating_sub(paper_x * 2);
    let paper_h = canvas_h.saturating_sub(paper_y * 2);

    let paper_pad_x = (canvas_w as f32 * 0.06).round() as u32;
    let paper_pad_top = (canvas_h as f32 * 0.06).round() as u32;
    let caption_h = if config.show_watermark {
        ((canvas_h as f32 * 0.14).round() as u32).max((76.0 * rs) as u32)
    } else {
        (canvas_h as f32 * 0.06).round() as u32
    };

    let area_x = paper_x + paper_pad_x;
    let area_y = paper_y + paper_pad_top;
    let area_w = paper_w.saturating_sub(paper_pad_x * 2);
    let area_h = paper_h.saturating_sub(paper_pad_top + caption_h);
    let (pw, ph, ox, oy) = fit_photo_in_area(area_w, area_h, src_w, src_h);
    let photo_x = (area_x + ox) as i64;
    let photo_y = (area_y + oy) as i64;
    let placed_x = area_x + ox;
    let placed_y = area_y + oy;

    let mut canvas = RgbaImage::from_pixel(canvas_w, canvas_h, PAPER_BLACK);
    fill_rect(&mut canvas, paper_x as i64, paper_y as i64, paper_w, paper_h, PAPER_WHITE);

    let photo = resize_photo(source, pw, ph);
    overlay(&mut canvas, &photo, photo_x, photo_y);

    if !config.show_watermark {
        return canvas;
    }

    let Some(rend) = text else { return canvas; };

    let spec = watermark_layout_spec();
    let line_pt = (frame.font_size as f32 * spec.secondary_font_scale).max(11.0 * rs);
    let caption_gap = (canvas_h as f32 * 0.025).round();
    let caption_top = placed_y as f32 + ph as f32 + caption_gap;

    // PROOF stamp: draw upright (no rotation - Rust image crate limitation)
    let stamp_pt = (canvas_w as f32 * 0.038).max(22.0 * rs);
    let stamp_y = caption_top + (canvas_h as f32 * 0.045).max(28.0 * rs);
    let stamp_text = "PROOF";
    let stamp_w = rend.measure(stamp_text, stamp_pt, true);
    let stamp_h = rend.line_height(stamp_pt, true);
    let box_pad_x = (6.0 * rs).round() as u32;
    let box_pad_y = (4.0 * rs).round() as u32;
    // Draw box outline
    let bx = placed_x as i64 - box_pad_x as i64;
    let by = (stamp_y - rend.ascent(stamp_pt, true) - box_pad_y as f32) as i64;
    let bw = stamp_w as u32 + box_pad_x * 2;
    let bh = stamp_h as u32 + box_pad_y * 2;
    draw_rect_outline(&mut canvas, bx, by, bw, bh, 2, PROOF_RED);
    rend.draw(&mut canvas, stamp_text, placed_x as f32, stamp_y - stamp_h, stamp_pt, true, PROOF_RED);

    // Right column: params, camera, date (typewriter style)
    let right_x = placed_x + pw;
    let camera = if config.show_camera { format_camera(exif) } else { String::new() };
    let lens = if config.show_lens { clean_display_text(&exif.lens) } else { String::new() };
    let cam_lens: String = [camera, lens].into_iter().filter(|s| !s.is_empty()).collect::<Vec<_>>().join("  \u{00B7}  ");

    let mut meta_lines: Vec<(String, bool)> = Vec::new();
    if config.show_params {
        let p = build_params_line(exif, "  ");
        if !p.is_empty() { meta_lines.push((p, true)); }
    }
    if !cam_lens.is_empty() { meta_lines.push((cam_lens, false)); }
    if config.show_date {
        let d = format_date(&exif.taken_at, &config.date_format);
        if !d.is_empty() { meta_lines.push((d, false)); }
    }
    for (i, (line, bold)) in meta_lines.iter().enumerate() {
        let y = caption_top + i as f32 * line_pt * 1.55;
        let w = rend.measure(line, line_pt, *bold);
        let x = right_x as f32 - w;
        let color = if i == 0 { PAPER_INK } else { PAPER_MUTED };
        rend.draw(&mut canvas, line, x, y, line_pt, *bold, color);
    }

    canvas
}

fn draw_rect_outline(canvas: &mut RgbaImage, x: i64, y: i64, w: u32, h: u32, thickness: u32, color: Rgba<u8>) {
    fill_rect(canvas, x, y, w, thickness, color);                          // top
    fill_rect(canvas, x, y + h as i64 - thickness as i64, w, thickness, color); // bottom
    fill_rect(canvas, x, y, thickness, h, color);                          // left
    fill_rect(canvas, x + w as i64 - thickness as i64, y, thickness, h, color); // right
}
